File: src/csharp/Intel/Iced.UnitTests/Intel/FormatterTests/NumberFileReader.cs
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

#if GAS || INTEL || MASM || NASM
using System;
using System.Collections.Generic;
using System.IO;

namespace Iced.UnitTests.Intel.FormatterTests {
	static class NumberFileReader {
		public static IEnumerable<object> ReadNumberFile(string filename) {
			int lineNo = 0;
			foreach (var line in File.ReadLines(FileUtils.GetFormatterFilename("Number"))) {
				lineNo++;
				if (line.Length == 0 || line[0] == '#')
					continue;
				object testCase;
				try {
					testCase = ReadTestCase(line, lineNo);
				}
				catch (Exception ex) {
					throw new InvalidOperationException($"Error parsing number test case file '{filename}', line {lineNo}: {ex.Message}");
				}
				yield return testCase;
			}
		}

		static readonly char[] seps = new char[] { ',' };
		static object ReadTestCase(string line, int lineNo) {
			var parts = line.Split(seps);
			if (parts.Length != 2)
				throw new InvalidOperationException($"Invalid number of commas ({parts.Length - 1} commas)");

			var valueStr = parts[1].Trim();
			switch (parts[0].Trim()) {
			case "i8": return NumberConverter.ToInt8(valueStr);
			case "u8": return NumberConverter.ToUInt8(valueStr);
			case "i16": return NumberConverter.ToInt16(valueStr);
			case "u16": return NumberConverter.ToUInt16(valueStr);
			case "i32": return NumberConverter.ToInt32(valueStr);
			case "u32": return NumberConverter.ToUInt32(valueStr);
			case "i64": return NumberConverter.ToInt64(valueStr);
			case "u64": return NumberConverter.ToUInt64(valueStr);
			default: throw new InvalidOperationException($"Invalid type: {parts[0]}");
			}
		}
	}
}
#endif

File: src/java/iced-x86/src/test/java/com/github/icedland/iced/x86/fmt/NumberFileReader.java
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

package com.github.icedland.iced.x86.fmt;

import java.util.ArrayList;

import com.github.icedland.iced.x86.FileUtils;
import com.github.icedland.iced.x86.NumberConverter;

final class NumberFileReader {
	public static TestNumber[] readNumberFile(String filename) {
		ArrayList<TestNumber> result = new ArrayList<TestNumber>();
		int lineNo = 0;
		for (String line : FileUtils.readAllLines(FmtFileUtils.getFormatterFilename("Number"))) {
			lineNo++;
			if (line.length() == 0 || line.charAt(0) == '#')
				continue;
			TestNumber testCase;
			try {
				testCase = readTestCase(line, lineNo);
			}
			catch (Exception ex) {
				throw new UnsupportedOperationException(
						String.format("Error parsing number test case file '%s', line %d: %s", filename, lineNo, ex.getMessage()));
			}
			result.add(testCase);
		}
		return result.toArray(new TestNumber[0]);
	}

	static TestNumber readTestCase(String line, int lineNo) {
		String[] parts = line.split(",");
		if (parts.length != 2)
			throw new UnsupportedOperationException(String.format("Invalid number of commas (%d commas)", parts.length - 1));

		String valueStr = parts[1].trim();
		switch (parts[0].trim()) {
		case "i8":
			return new TestNumber(TestNumberKind.BYTE, NumberConverter.toInt8(valueStr));
		case "u8":
			return new TestNumber(TestNumberKind.UBYTE, NumberConverter.toUInt8(valueStr));
		case "i16":
			return new TestNumber(TestNumberKind.SHORT, NumberConverter.toInt16(valueStr));
		case "u16":
			return new TestNumber(TestNumberKind.USHORT, NumberConverter.toUInt16(valueStr));
		case "i32":
			return new TestNumber(TestNumberKind.INT, NumberConverter.toInt32(valueStr));
		case "u32":
			return new TestNumber(TestNumberKind.UINT, NumberConverter.toUInt32(valueStr));
		case "i64":
			return new TestNumber(TestNumberKind.LONG, NumberConverter.toInt64(valueStr));
		case "u64":
			return new TestNumber(TestNumberKind.ULONG, NumberConverter.toUInt64(valueStr));
		default:
			throw new UnsupportedOperationException(String.format("Invalid type: %s", parts[0]));
		}
	}
}
